llama_index/node_parser/unstructured_element.py

```diff
@@ -0,0 +1,303 @@
+            # flush text buffer
+            if len(cur_text_el_buffer) > 0:
+                cur_text_nodes = _get_nodes_from_buffer(cur_text_el_buffer, node_parser)
+                nodes.extend(cur_text_nodes)
+                cur_text_el_buffer = []
+
+            table_output = cast(TableOutput, element.table_output)
+            table_df = cast(pd.DataFrame, element.table)
+
+            table_id = element.id + "_table"
+            table_ref_id = element.id + "_table_ref"
+            # TODO: figure out what to do with columns
+            # NOTE: right now they're excluded from embedding
+            col_schema = "\n\n".join([str(col) for col in table_output.columns])
+            index_node = IndexNode(
+                text=str(table_output.summary),
+                metadata={"col_schema": col_schema},
+                excluded_embed_metadata_keys=["col_schema"],
+                id_=table_ref_id,
+                index_id=table_id,
+            )
+            table_str = table_df.to_string()
+            text_node = TextNode(
+                text=table_str,
+                id_=table_id,
+            )
+            nodes.extend([index_node, text_node])
+        else:
+            cur_text_el_buffer.append(str(element.element))
+
+    # flush text buffer
+    if len(cur_text_el_buffer) > 0:
+        cur_text_nodes = _get_nodes_from_buffer(cur_text_el_buffer, node_parser)
+        nodes.extend(cur_text_nodes)
+        cur_text_el_buffer = []
+
+    return nodes
+
+
+def get_base_nodes_and_mappings(nodes: List[BaseNode]) -> Tuple[List[BaseNode], Dict]:
+    """Get base nodes and mappings.
+
+    Givne a list of nodes and IndexNode objects, return the base nodes and a mapping
+    from index id to child nodes (which are excluded from the base nodes).
+
+    """
+    node_dict = {node.node_id: node for node in nodes}
+
+    node_mappings = {}
+    base_nodes = []
+
+    # first map index nodes to their child nodes
+    nonbase_node_ids = set()
+    for node in nodes:
+        if isinstance(node, IndexNode):
+            node_mappings[node.index_id] = node_dict[node.index_id]
+            nonbase_node_ids.add(node.index_id)
+        else:
+            pass
+
+    # then add all nodes that are not children of index nodes
+    for node in nodes:
+        if node.node_id not in nonbase_node_ids:
+            base_nodes.append(node)
+
+    return base_nodes, node_mappings
+
+
+DEFAULT_SUMMARY_QUERY_STR = """\
+What is this table about? Give a very concise summary (imagine you are adding a caption), \
+and also output whether or not the table should be kept.\
+"""
+
+
+class UnstructuredElementNodeParser(NodeParser):
+    """Unstructured element node parser.
+
+    Splits a document into Text Nodes and Index Nodes corresponding to embedded objects
+    (e.g. tables).
+
+    """
+
+    callback_manager: CallbackManager = Field(
+        default_factory=CallbackManager, exclude=True
+    )
+    llm: Optional[Any] = Field(
```

Comment:
Should this be `Optional[LLM]` ?

Reply:
i was going to make an argument about circular deps but realized we already top-level import OpenAI, so i'll just type this as LLM